enhancements/storage/azure-file-cloning-with-azcopy.md

```diff
@@ -0,0 +1,202 @@
+
+### Workflow Description
+
+Clones are provisioned like any other PVC with the exception of adding a dataSource that references an existing PVC in the same namespace.
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+    name: clone-of-pvc-1
+    namespace: myns
+spec:
+  accessModes:
+  - ReadWriteOnce
+  storageClassName: cloning
+  resources:
+    requests:
+      storage: 5Gi
+  dataSource:
+    kind: PersistentVolumeClaim
+    name: pvc-1
+```
+
+The result is a new PVC with the name clone-of-pvc-1 that has the exact same content as the specified source pvc-1
+
+#### Variation and form factor considerations [optional]
+
+None.
+
+### API Extensions
+
+None.
+
+### Implementation Details/Notes/Constraints [optional]
+
+None.
+
+#### Hypershift [optional]
+
+None.
+
+### Risks and Mitigations
+
+Cloning is a standard feature of the CSI spec and is supported by most CSI drivers. The only risk could be that the
+performance of the cloning operation is poor for larger files, we intend to test this to mitigate the risk.
+
+### Drawbacks
+
+None.
+
+## Design Details
+
+We need to get `azcopy` into the container with the Azure File CSI Driver. The `azcopy` tool is a tool written in Go that
+supports many different operations apart from copying objects, but we don't intend to support any if the extra features
+apart from those required by cloning.
+
+We've chosen to use base image approach due to the following reasons:
+- we're familiar with the process
+- we don't want to support a generic RPM that can be used outside of OCP
+- we want to keep consistency with other tools (e.g. AWS EFS CSI Driver)
+
+1. Fork `github.com/Azure/azure-storage-azcopy` into `github.com/openshift/azure-storage-azcopy`.
+
+2. Create a new base image with `azcopy` included, say `ose-azure-storage-azcopy-base`:
+
+   ```Dockerfile
+   FROM registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.20-openshift-4.16 AS builder
+   WORKDIR /go/src/github.com/openshift/azure-storage-azcopy
+   COPY . .
+   RUN go build -o ./bin/azcopy .
+    
+   FROM registry.ci.openshift.org/ocp/4.16:base
+   COPY --from=builder /go/src/github.com/openshift/azure-storage-azcopy/bin/azcopy /usr/bin/
+   ```
+3. The Azure File CSI Driver then uses it as the base image:
+  
+   ```Dockerfile
+   FROM registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.20-openshift-4.16 AS builder
+   ... build the image ...
+   FROM registry.ci.openshift.org/ocp/4.16:ose-azure-storage-azcopy-base
+   COPY --from=builder <driver binary> /usr/bin/
+   ```
+
+### Open Questions [optional]
+
+* How do we determine what performance or stress testing is required for this feature? We should assess what volume
+sizes are the most common and test against those.
+
+* According to upstream only samba is supported for cloning, should we support and test only samba or do we want NFS support as well?
+https://github.com/kubernetes-sigs/azurefile-csi-driver/tree/master/deploy/example/cloning#feature-status-beta
```

Comment:
We must make sure that the driver returns a reasonable error when cloning a NFS volume.

Reply:
Ack, moving to goals.